pkg/agent/endpoints/sdsv3/handler.go

```diff
@@ -507,14 +508,38 @@ func supportsSPIFFEAuthExtension(req *discovery_v3.DiscoveryRequest) bool {
 }
 
 func (h *Handler) isSPIFFECertValidationDisabled(req *discovery_v3.DiscoveryRequest) bool {
-	if h.c.DisableSPIFFECertValidation {
+	fields := req.Node.GetMetadata().GetFields()
+	v, ok := fields[disableSPIFFECertValidationKey]
+	var vb bool
+	if ok {
+		var err error
+		vb, err = parseBool(v)
+		if err != nil {
+			// error means that value isn't parsable string
+			// so it would be safer to assume that key doesn't exist in envoy node metadata
+			ok = false
+		}
+	}
+
+	if h.c.DisableSPIFFECertValidation && (!ok || vb) {
 		return true
 	}
 
-	fields := req.Node.GetMetadata().GetFields()
-	v, ok := fields[disableSPIFFECertValidationKey]
+	return vb
+}
+
+func parseBool(v *structpb.Value) (bool, error) {
+	str := v.GetStringValue()
+	if str != "" {
+		strBool, err := strconv.ParseBool(str)
+		if err != nil {
+			return false, err
+		}
+
+		return strBool, nil
+	}
 
-	return ok && (v.GetBoolValue() || v.GetStringValue() == "true")
+	return v.GetBoolValue(), nil
```

Comment:
I think we need to be a little more stringent here. As written, if somebody put in a float value or something, it would override the default disabled value. I think we should reject anything but a bool or a string:

Something like:

```
func parseBool(v *structpb.Value) (bool, error) {
	switch v := v.GetKind().(type) {
	case *structpb.Value_BoolValue:
		return v.BoolValue, nil
	case *structpb.Value_StringValue:
		return strconv.ParseBool(v.StringValue)
	default:
		return false, fmt.Errorf("unsupported value type %T", v)
	}
}
```

Reply:
nice catch. will fix it